rust/connlib/clients/shared/src/lib.rs

```diff
@@ -139,27 +133,18 @@ where
     fn connect_inner(
         runtime: &Runtime,
         runtime_stopper: tokio::sync::mpsc::Sender<StopRuntime>,
-        api_url: Url,
-        token: SecretString,
-        device_id: String,
-        device_name_override: Option<String>,
+        url: LoginUrl,
         os_version_override: Option<String>,
         callbacks: CallbackErrorFacade<CB>,
         max_partition_time: Option<Duration>,
     ) {
         runtime.spawn(async move {
-            let (connect_url, private_key) = fatal_error!(
-                login_url(Mode::Client, api_url, token, device_id, device_name_override),
-                runtime_stopper,
-                &callbacks
-            );
-
             // This is kinda hacky, the buffer size is 1 so that we make sure that we
             // process one message at a time, blocking if a previous message haven't been processed
             // to force queue ordering.
             let (control_plane_sender, mut control_plane_receiver) = tokio::sync::mpsc::channel(1);
 
-            let mut connection = PhoenixChannel::<_, IngressMessages, ReplyMessages, Messages>::new(Secret::new(SecureUrl::from_url(connect_url)), os_version_override, move |msg, reference, topic| {
+            let mut connection = PhoenixChannel::<_, IngressMessages, ReplyMessages, Messages>::new(Secret::new(SecureUrl::from_url(Url::parse(url.url().expose_secret().as_ref()).expect("we know it is a valid URL"))), os_version_override, move |msg, reference, topic| {
```

Comment:
This is a bit ugly. I am going to send another PR that cleans this up first and then rebase this one on top.